### ISSUE TITLE
Multicast and LogReceiverTarget

### ISSUE THREAD

**Issue by MartinTherriault** [NLog/NLog#561]
I want to use the LogReceiverTarget with a WCF router service to be able to dynamically multicast the logs message.

One of the requirement for multicast is that the ServiceContract must only allow one way messages.
Right now the ILogReceiverServer contract does not meets the requirement because its only method (ProcessLogMessages) is not one way, but it already has a void return type.

Would it be possible to change the OperationContract attribute of said method to add "IsOneWay=true"? I already have tested this change and it works fine in my multicast scenario.

Otherwise I'll just re-implement it all in my own custom target if there would be any possible issue with this change.

Thanks

**Comment by 304NotModified** [comment]
A Pull Request will be accepted for this!

**Comment by 304NotModified** [comment]
PR is merged so closing this.

**Comment by 304NotModified** [comment]
Hi @MartinTherriault

We are reopening this issue because we having some problems with the merged PR.

- It's not backwards-compatible - but we thought it was. (see #565)
- We are missing a lot of properties in the `WcfLogReceiverClientFacade`
- We have a lot of code duplication.

I'm ATM busy writing some good unit tests for this (with really starting a WCF server). After that, I think we should refactor a lot to fix the issues. Can you help with that?

**Comment by MartinTherriault** [comment]
Of course I will help. I looked at the issues and now believe the facade may not be the right approach. I think there would be two easy ways to fix it. In both case we would revert the changes.
The first way would be to add the new one way method to the service contract (I think it's not breaking, but would have to confirm) and send the log message on the right contract method specified in the target (one or two way).
The second way would be to create a new target for this and reuse most of the code already in place.


Date: Thu, 2 Jul 2015 14:31:23 -0700

**Comment by 304NotModified** [comment]
Thanks for the response!

About the possible solutions:

new one way method to current service contract
==
I was also thinking about this. But if I understand correct: we use another methodname for the oneway contract, and we have 2 interfaces? Because, if we have 1 interface, then this change would we a breaking change.

new target
==
I think we should look a this solution if the other isn't working (because of breaking changes for example) - because both targets would have almost the same functionality and should be the same class for an architectural point of view. 

I did also checked if we could change the WCF properties with an API instead of attributes, but that seems not be possible to bad.

**Comment by MartinTherriault** [comment]
Something just hit me this morning, I've been thinking of this backward. NLog is a WCF client, so we cannot just add a method to the contract since the server side would no longer compile without implementing the new method.

We have to inherit the service contract in a new interface (named V2 for quick reference) and add the new method there. If the server side wants the new method, it has to use the V2 contract and implement the new method. For the client side (NLog), we implement the V2 contract and we add a property to the target to select which method to use to send events, default being the original two way method.

All this would be nonbreaking. Of course if someone was to set the target to use the V2 method on a server implementing the original contract, it would fail. 


For what its worth, I did a quick search this morning before realizing my backward thinking mistake and it seems adding a method to the service contract is nonbreaking as I thought.

From MSDN:

>Contracts used by a client do not need to be the same as the contract used by the service; they need only to be compatible.
>For service contracts, compatibility means new operations exposed by the service can be added but existing operations cannot be removed or changed semantically.

and
 
>Adding service operations exposed by the service is a nonbreaking change because existing clients need not be concerned about those new operations.

**Comment by 304NotModified** [comment]
> Something just hit me this morning, I've been thinking of this backward. NLog is a WCF client, so we cannot just add a method to the contract since the server side would no longer compile without implementing the new method.

Yes, I was trying to say that before ;)

> We have to inherit the service contract in a new interface (named V2 for quick reference) and add the new method there. If the server side wants the new method, it has to use the V2 contract and implement the new method. For the client side (NLog), we implement the V2 contract and we add a property to the target to select which method to use to send events, default being the original two way method.

Can't we have two interfaces and one implementing client? (an user could choose witch interface(s) he would implement)



>From MSDN:

>>Contracts used by a client do not need to be the same as the contract used by the service; they need only to be compatible.
>>For service contracts, compatibility means new operations exposed by the service can be added but existing operations cannot be removed or changed semantically.

>and
 
>>Adding service operations exposed by the service is a nonbreaking change because existing clients need not be concerned about those new operations.

I'm not sure what we should conclude out of this. Can you summarize this?

**Comment by 304NotModified** [comment]
It would be nice if we could fix this within 2 weeks and include this in 4.1.

Let continue the discussion at https://github.com/NLog/NLog/issues/784. A new ticket for this problem is more suitable.
